NativePromise::all should handle NativePromise with ResolveValueT==void

jyavenard opened the pull request:
#### cc6dee38764b3229258d2dba36fe140ecd942bf8
<pre>
NativePromise::all should handle NativePromise with ResolveValueT==void
<a href="https://bugs.webkit.org/show_bug.cgi?id=264082">https://bugs.webkit.org/show_bug.cgi?id=264082</a>
<a href="https://rdar.apple.com/117837167">rdar://117837167</a>

Reviewed by Youenn Fablet.

Also fix bug 264122 as fly-by

Added API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270344@main">https://commits.webkit.org/270344@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e73ac74ec55d5b61dc32843a31cc6c22fdbdea7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27290 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23099 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25444 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1152 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23341 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27869 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2427 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22667 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28791 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21892 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26617 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24413 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/672 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31838 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3726 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6971 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6047 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2816 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2711 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -429,7 +429,7 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         return p;
```
Nit about the commit description:
"NativePromise::all can't handle NativePromise with ResolveValueT==void"
This is describing the problem that you're resolving (so it's the situation before the patch), it could be misread as "NP::all *shouldn't* handle void".
For me, when looking at commit summaries, I like to know what patches *do* in a direct manner, e.g. for this one: "NativePromise::all should handle NativePromise with ResolveValueT==void" or "Add support for void ResolveValueT in NativePromise::all".
Then in the commit description details, you could optionally add how things were before the patch if it's not obvious, expand what the goal is, and how the patch is achieving it.

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -546,13 +552,14 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
 
     private:
-        Vector<std::optional<ResolveValueType>> m_resolveValues;
+        NO_UNIQUE_ADDRESS std::conditional_t<!std::is_void_v<ResolveValueType>, Vector<std::optional<ResolveValueType>>, std::monostate> m_resolveValues;
```
`std::monostate` is declared in `<variant>`, you should include it to avoid potential broken builds.
(I've just noticed it was already used in this file, missed that in previous reviews.)

Optional alternative: It feels a bit like an abuse of that type that is intended for use in `std::variant`.
Since they're not publicly visible here, would it be possible to replace them with private empty types?

jyavenard replied:
std::monostate seems to have been used in combination of NO_UNIQUE_ADDRESS to generate a 0-bytes member. I only followed existing examples.

I'm not sure on which to use otherwise.

squelart replied:
I was suggesting you create your own empty type right above this line. 😉
Something like:
```
struct VoidPlaceholder {};
NO_UNIQUE_ADDRESS std::conditional_t<!std::is_void_v<ResolveValueType>, Vector<std::optional<ResolveValueType>>, VoidPlaceholder> m_resolveValues;
```

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -593,27 +600,33 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-                else
+                if (result) {
+                    if constexpr (std::is_void_v<ResolveValueT>)
+                        producer->resolve(i, std::nullptr_t());
```
I was not sure why you were using `nullptr_t()` here, I had to look at `resolve()` to understand that it's not actually used, so it could be anything.
To reduce this potential confusion, I would suggest you use the same `std::monostate` (or other empty struct as suggested above), it will be easier to make the connection with your void-type-alternative.

jyavenard replied:
yeah, I only used it to avoid duplicating code as I needed an argument, any argument.

squelart reviewed (approved):
The code looks correct, so of course I had to pick nits about less-important things. 😄

youennf commented on Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
```diff
@@ -699,6 +699,24 @@ TEST(NativePromise, PromiseAllResolve)
     });
 }
 
+TEST(NativePromise, PromiseVoidAllResolve)
```
Can you add a test for Promise.all with r-values?

youennf commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -593,27 +600,33 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
-
+        if (promises.isEmpty()) {
+            if constexpr (std::is_void_v<ResolveValueT>)
+                return AllPromiseType::createAndResolve();
```
I wonder whether we could avoid some of these constexpr if checks by having AllPromiseType::ResolveValueType be std::nullptr_t in case of void.
As mentioned by @squelart, we are sometimes using nullptr for void and sometimes special casing things.

jyavenard replied:
no, nullptr is only used internally to pass an (unused) argument to a method expecting one. It's not used outside of NativePromise internal

jyavenard replied:
Returning a `Vector<nullptr_t>` seems weird to me, I prefer having AllPromise::ResolveValueType to be `void`

jyavenard committed:
NativePromise::all should handle NativePromise with ResolveValueT==void
https://bugs.webkit.org/show_bug.cgi?id=264082
rdar://117837167

Reviewed by Youenn Fablet.

Also fix bug 264122 as fly-by

Added API tests.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270344@main
